src/lib/common/include/sol-mainloop.h

```diff
@@ -64,11 +64,269 @@ extern "C" {
+ *    </ul>
+ * <li> @c rx_cb: The read callback - It will be called when there's data available to read.
+ *    <ul>
+ *    <li> Respecting the interface:
+ *    @code{.c}
+ *    ssize_t (*rx_cb)(void *user_data, my_stream_api_handle *handle, const struct sol_buffer *buf);
+ *    @endcode
+ *    </ul>
+ * <li> @c user_data: Data to @c rx_cb and @c tx_cb.
+ * </ul>
+ *
+ * Below there's a example of a configuration struct.
+ * @code{.c}
+ *
+ * //my_stream_api.h
+ *
+ * struct my_stream_api_config {
+ *    const void *user_data;
+ *    void (*tx_cb)(void *user_data, my_stream_api_handle *handle, struct sol_blob *blob, int status);
+ *    ssize_t (*rx_cb)()void *user_data, my_stream_api_handle *handle, const struct sol_buffer *buf);
+ *    size_t tx_size;
+ *    size_t rx_size;
+ * };
+ *
+ * //Code...
+ *
+ * //The config struct should be provided during the stream creation.
+ * struct my_stream_api_handle *my_stream_api_new(struct my_stream_api_config *config);
+ *
+ * //More code...
+ * @endcode
+ *
+ * @subsection write Writing
+ *
+ * Write operations must be async and the data that will be transfered must be
+ * provided as a pointer to @ref sol_blob. One should follow the API below:
+ * @code{.c}
+ * //returns 0 on success, -ENOSPC if tx_buffer is FULL or negative errno.
+ * int my_stream_api_write(struct my_stream_api_handle *handle, struct sol_blob *blob);
+ * @endcode
+ *
+ * Every blob requested to be written, must be queued in order to be sent. If there's no more space
+ * to queue more blobs (sum of all queued blobs >= @c tx_size) the function write should return @c -ENOSPC.
+ *
+ * Everytime a blob is fully written the @c tx_cb must be called in order to inform the user that
+ * the write operation was completed. The @c tx_cb should have the following signature:
+ *
+ * @code{.c}
+ * void (*tx_cb)(void *user_data, my_stream_api_handle *handle, struct sol_blob *blob, int status);
+ * @endcode
+ *
+ * One should warn the user that it's not necessary to use @ref sol_blob_unref(), because the stream
+ * API will already do that. The @c status variable should be 0 on success and @c -errno on error.
+ * The @c tx_cb can also be used to send blobs that were reject, because the tx buffer was full due
+ * @c tx_size restrictions.
+ *
+ * Below there's an example of a write implementation:
+ * @code{.c}
+ *
+ * //my_stream_api.c
+ *
+ * struct my_stream_api_handle {
+ *    Device *dev;
+ *    const void *user_data;
+ *    void (*tx_cb)(void *user_data, my_stream_api_handle *handle, struct sol_blob *blob, int status);
+ *    struct sol_ptr_vector pending_blobs;
+ *    size_t tx_size;
+ *    size_t pending_bytes;
+ * };
+ *
+ *
+ * static void
+ * _write(void *data)
+ * {
+ *    struct sol_blob *blob;
+ *    struct my_stream_api_handle *handle = data;
+ *    int status;
+ *
+ *    //Write the blob
+ *    blob = sol_ptr_vector_get_no_check(&handle->pending_blobs, 0);
+ *    status = my_device_write(handle->dev, blob->mem, blob->size);
```

Comment:
go with write(2) like `my_device_write()` signature, and show in the example that the device may return less than proposed bytes, so we document the pending behavior